runners/portability/java/src/main/java/org/apache/beam/runners/portability/ExternalWorkerService.java

```diff
@@ -90,10 +97,55 @@ public void stopWorker(
   public void close() {}
 
   public GrpcFnServer<ExternalWorkerService> start() throws Exception {
-    GrpcFnServer<ExternalWorkerService> server =
-        GrpcFnServer.allocatePortAndCreateFor(this, serverFactory);
+    final String externalServiceAddress =
+        Environments.getExternalServiceAddress(options.as(PortablePipelineOptions.class));
+    GrpcFnServer<ExternalWorkerService> server;
+    if (externalServiceAddress.isEmpty()) {
+      server = GrpcFnServer.allocatePortAndCreateFor(this, serverFactory);
+    } else {
+      server =
+          GrpcFnServer.create(
+              this,
+              Endpoints.ApiServiceDescriptor.newBuilder().setUrl(externalServiceAddress).build(),
+              serverFactory);
+    }
     LOG.debug(
         "Listening for worker start requests at {}.", server.getApiServiceDescriptor().getUrl());
     return server;
   }
+
+  /**
+   * Worker pool entry point.
+   *
+   * <p>The worker pool exposes an RPC service that is used with EXTERNAL environment to start and
+   * stop the SDK workers.
+   *
+   * <p>The worker pool uses threads for parallelism;
+   *
+   * <p>This entry point is used by the Java SDK container in worker pool mode.
+   */
+  public static void main(String[] args) throws Exception {
+    main(System::getenv);
+  }
+
+  public static void main(Function<String, String> environmentVarGetter) throws Exception {
+    System.out.format("Starting external worker service%n");
+    System.out.format("Pipeline options %s%n", environmentVarGetter.apply(PIPELINE_OPTIONS));
```

Comment:
This code assumes pipeline options are known ahead of time. Ideally, a worker pool should be logically uncoupled from a single pipeline, and able to serve multiple pipelines with varying options. It may not be possible, however, without modifying StartWorkerRequest to include pipeline options.

This wasn't a problem in Python. The difference is that in Java, FnHarness takes pipeline options as an argument. sdk_worker.py, as far as I can tell, does not use pipeline options.

It may be acceptable to take pipeline options up-front, but it should be clear from documentation that   `PIPELINE_OPTIONS` should match the pipeline options used by the job, or else there could be confusion about which set of options is actually read in various places.

Reply:
Completely agree!

I noticed this difference too, however, it is not straightforward to exclude pipeline options when running `FnHarness`. Modifying StartWorkerRequest to include pipeline options could help decouple the requirement which should  be addressed in a separate PR if we decided this is needed.

I have updated the javadoc further to clearly pointed out this restriction.